File: app/src/main/java/com/hwonchul/movie/data/local/dao/VideoDao.java
package com.hwonchul.movie.data.local.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.hwonchul.movie.data.local.model.VideoEntity;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;

@Dao
public interface VideoDao {

    @Query("SELECT * FROM " + VideoEntity.TABLE_NAME + " WHERE movie_id = :movieId " +
            "ORDER BY published_at DESC")
    Flowable<List<VideoEntity>> findVideosByMovieId(int movieId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Completable insert(List<VideoEntity> entities);
}

File: app/src/main/java/com/hwonchul/movie/data/local/dao/VideoDao.kt
package com.hwonchul.movie.data.local.dao

import android.database.sqlite.SQLiteConstraintException
import androidx.room.Dao
import androidx.room.Query
import com.hwonchul.movie.base.data.dao.BaseDao
import com.hwonchul.movie.data.local.model.VideoEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface VideoDao : BaseDao<VideoEntity> {
    @Query(
        "SELECT * FROM " + VideoEntity.TABLE_NAME + " WHERE movie_id = :movieId " +
                "ORDER BY published_at DESC"
    )
    fun findVideosByMovieId(movieId: Int): Flow<List<VideoEntity>>

    suspend fun upsert(entity: VideoEntity) {
        try {
            insert(entity)
        } catch (e: SQLiteConstraintException) {
            update(entity)
        }
    }

    suspend fun upsert(entities: List<VideoEntity>) {
        entities.forEach { entity ->
            try {
                insert(entity)
            } catch (e: SQLiteConstraintException) {
                update(entity)
            }
        }
    }
}